### PR TITLE
[SWP-2832] Auth Service Upgrade: 5.6+ Package

### PR DESCRIPTION
**Pull request by HasnatH**
### Fix ?
Link to ticket: https://podpoint.atlassian.net/browse/SWP-2832

### What has been done
- update structure so it can be added as a logging channel
- supports Laravel 5.6+

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
     "type": "library",
     "require": {
         "php": ">=7.0",
-        "illuminate/support": "^5.6",
+        "illuminate/support": "^5.7|^6.0|^7.0|^8.0",
```
Comment (willthorpe) [inline composer.json]: Looking into this more from Laravel 7 onwards support for monolog 1 was dropped. https://github.com/laravel/framework/pull/29866. We will need another separate release for Laravel 7/8 using Monolog 2 otherwise we can't install dependencies from Laravel 7+

File: tests/unit/ServiceProviderTest.php [hunk 2]
```diff
@@ -19,6 +19,24 @@ protected function getEnvironmentSetUp($app)
...
+                    'level' => config('kinesis.level'),
+                ],
+                'formatter' => \PodPoint\KinesisLogger\Monolog\KinesisFormatter::class,
+                'formatter_with' => [
+                    'name' => config('app.name'),
+                ],
+            ],
+        ]);
+
+        $app['config']->set('logging.default', 'kinesis');
```
Comment (willthorpe) [inline tests/unit/ServiceProviderTest.php]: Looking at this with Andrew we found that the instructions in this package show that we will be using config file data inside other config files - for example kinesis.stream and kinesis.level data is relied on inside the config/logging file. This can cause an issue depending on which config files are loaded first. 

A solution can be to get rid of kinesis.php moving all necessary data within the kinesis object and then removing the config dependencies. 

'formatter_with' => [
        'name' => config('app.name'),
    ],

can become:
'formatter_with' => [
        'name' => env('app.name') ?? 'Laravel',
    ],

I think @oaklees sent more information about a different solution for this.

**Review by willthorpe — DISMISSED** [review]
Couple of things spotted with @oaklees 

Also in KinesisFormatterTest
We should also test that the record array matches the output from the formatter. 
$this->assertArrayHasKey('Data', $output);

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,6 @@
 composer.phar
 /vendor/
+build
```
Comment (oaklees) [inline .gitignore:4]: ?

File: README.md [hunk 4]
```diff
@@ -22,14 +22,27 @@ Add the service provider to your `config/app.php` providers array:
 ]
 ```
 
-Then finally, publish the config files:
```
Comment (oaklees) [inline README.md:25]: Our composer.json has the Laravel autodiscovery fields, so don't need to manually specify ^^ in our `app.php`

File: composer.lock [hunk 5]
```diff
@@ -0,0 +1,5971 @@
+{
```
Comment (oaklees) [inline composer.lock]: Can we remove this file

File: composer.json [hunk 6]
```diff
@@ -5,7 +5,7 @@
     "type": "library",
     "require": {
         "php": ">=7.0",
```
Comment (oaklees) [inline composer.json]: If we're changing our minimum Laravel version this should be incremented too

File: composer.json (hunk 1 above)
Comment (oaklees) [reply composer.json]: From discussions, to prevent having to support multiple releases, we are only going to be supporting L6+ moving forward (which is the earliest that support Monolog 2) so need to make sure we use Monolog 2 instead and test this still works as expected. See https://github.com/laravel/framework/blob/43518afa5253600513b19ba7c3d4aad2a3eb69c3/composer.json#L27.

We may also need to update our version of testbench to support this.

File: tests/unit/ServiceProviderTest.php [hunk 7]
```diff
@@ -19,6 +19,24 @@ protected function getEnvironmentSetUp($app)
...
+                'with' => [
+                    'streamName' => config('kinesis.stream'),
+                    'level' => config('kinesis.level'),
+                ],
+                'formatter' => \PodPoint\KinesisLogger\Monolog\KinesisFormatter::class,
+                'formatter_with' => [
+                    'name' => config('app.name'),
+                ],
+            ],
+        ]);
```
Comment (oaklees) [inline tests/unit/ServiceProviderTest.php]: Can we rewrite these to be reflective of the new channel method of configuring logs. Ideally we should be testing this with all major support Laravel versions. See https://github.com/Pod-Point/countries/blob/master/.github/workflows/run-tests.yml on how to do this with Actions and the matrix strategy.
Also, can we add more meaningful tests? As a minimum we should be able to mock the Kinesis client, setup the channel, send a log, and assert the client gets `putRecord` called with the expected body. We should do this for Laravel 6, 7 & 8 with both versions of PHP.

File: src/Providers/ServiceProvider.php [hunk 8]
```diff
@@ -22,9 +24,19 @@ public function boot()
...
+        }
+
+        app()->bind(KinesisFormatter::class, function () {
+            return new KinesisFormatter(config('app.name'), $this->app->environment());
+        });
+
+        app()->bind(KinesisClient::class, function () {
             $config = [
                 'region' => config('kinesis.aws.region'),
                 'version' => 'latest'
```
Comment (oaklees) [inline src/Providers/ServiceProvider.php]: Can we update the `config/kinesis.php` to only have the AWS crednetials as we no longer specify the stream or log level in there but directly within the `logging.php` of the consuming application.

**Review by oaklees — CHANGES_REQUESTED** [review]
Getting there! Some updates needed though - config still contains params that we've moved to `logging.php`, and tests are pretty thin on the ground and not really doing much. As we're moving to only supporting 6+ moving forward we need to move to monolog 2 - we should also as a result really be increasing this test coverage and testing against these major versions with both PHP versions. See notes below to the type of test we can do.

File: .styleci.yml [hunk 9]
```diff
@@ -0,0 +1 @@
+preset: laravel
```
Comment (oaklees) [inline .styleci.yml:1]: I've activated StyleCI - if you login with your GitHub account you should now be able to see the failures for this repo - please could you address them: https://github.styleci.io/analyses/EAlpYy

File: CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# Changelog
+
+All notable changes to `laravel-monolog-kinesis` will be documented in this file.
+
+## 3.0.0 - 2021-XX-XX
+
+- First release supporting PHP7.2+ and Laravel 6+
+- Support for Monolog 1.* and PHPUnit 7.* dropped.
```
Comment (oaklees) [inline CHANGELOG.md]: Probably clearer to say `Drop support for..`

File: CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,12 @@
+# Changelog
+
+All notable changes to `laravel-monolog-kinesis` will be documented in this file.
+
+## 3.0.0 - 2021-XX-XX
+
+- First release supporting PHP7.2+ and Laravel 6+
+- Support for Monolog 1.* and PHPUnit 7.* dropped. 
+- Switch to Github Actions from Travis CI.
```
Comment (oaklees) [inline CHANGELOG.md:9]: Can we also add that we've moved to the Laravel custom log driver implementation

File: README.md [hunk 12]
```diff
@@ -1,35 +1,71 @@
...
 ## Installation
 
 Require the package in composer:
 
+Laravel < 6.0
 ```javascript
 "require": {
     "pod-point/laravel-monolog-kinesis": "^2.0"
 },
 ```
```
Comment (oaklees) [inline README.md]: Can we change this to use the `composer require` syntax - no-one really manually updates composer.json these days. See https://github.com/Pod-Point/laravel-mail-export/blob/master/README.md

File: README.md [hunk 13]
```diff
@@ -1,35 +1,71 @@
...
 
+Laravel 6.0+
+```javascript
+"require": {
+    "pod-point/laravel-monolog-kinesis": "^3.0"
+},
+```
+
```
Comment (oaklees) [inline README.md]: Same as above

File: README.md [hunk 14]
```diff
@@ -1,35 +1,71 @@
...
 Add the service provider to your `config/app.php` providers array:
 
 ```php
 'providers' => [
-    PodPoint\KinesisLogger\Providers\ServiceProvider::class
+    PodPoint\KinesisLogger\KinesisMonologServiceProvider::class
 ]
 ```
```
Comment (oaklees) [inline README.md]: I don't believe we need this with the auto-discovery?

File: README.md [hunk 15]
```diff
@@ -1,35 +1,71 @@
...
 
-Then finally, publish the config files:
+Then, publish the config files:
 
 ```php
-php artisan vendor:publish --provider="PodPoint\KinesisLogger\Providers\ServiceProvider"
+php artisan vendor:publish --provider="PodPoint\KinesisLogger\KinesisMonologServiceProvider"
 ```
```
Comment (oaklees) [inline README.md:28]: Can we add this section under a new heading 'Usage'

File: phpunit.xml [hunk 16]
```diff
@@ -1,17 +1,27 @@
...
+<!--    <filter>-->
+<!--        <whitelist>-->
+<!--            <directory suffix=".php">src</directory>-->
+<!--            <exclude>-->
+<!--                <directory>vendor</directory>-->
+<!--                <directory>tests</directory>-->
+<!--            </exclude>-->
+<!--        </whitelist>-->
+<!--    </filter>-->
 </phpunit>
```
Comment (oaklees) [inline phpunit.xml]: ?

File: src/KinesisMonologServiceProvider.php [hunk 17]
```diff
@@ -19,12 +21,22 @@ class ServiceProvider extends LaravelServiceProvider
     public function boot()
     {
         $this->publishes([
-            __DIR__.'/../../config/kinesis.php' => config_path('kinesis.php'),
+            __DIR__ . '/../../config/kinesis.php' => config_path('kinesis.php'),
```
Comment (oaklees) [inline src/KinesisMonologServiceProvider.php]: Sorry forgot to mention - but we decided we should probably change the config file name to `monolog-kinesis` to prevent any conflicts as `kinesis` is quite generic and could easily reference many things.

File: src/KinesisMonologServiceProvider.php [hunk 18]
```diff
@@ -19,12 +21,22 @@ class ServiceProvider extends LaravelServiceProvider
...
+            });
+        }
+
+        app()->singleton(KinesisFormatter::class, function () {
+            return new KinesisFormatter(config('app.name'), $this->app->environment());
+        });
+
+        app()->singleton(KinesisClient::class, function () {
             $config = [
                 'region' => config('kinesis.aws.region'),
```
Comment (oaklees) [inline src/KinesisMonologServiceProvider.php]: Obviously need to update the `config` key when we change the file name

File: src/KinesisMonologServiceProvider.php [hunk 19]
```diff
@@ -19,12 +21,22 @@ class ServiceProvider extends LaravelServiceProvider
...
-            $monolog = Log::getLogger();
+        if ($this->app['log'] instanceof LogManager) {
+            $this->app['log']->extend('kinesis', function (Container $app, array $config) {
+                $handler = new KinesisHandler($config['stream'], $config['level'] ?? LogLevel::INFO);
 
+                return new Logger('kinesis', [$handler]);
+            });
+        }
+
+        app()->singleton(KinesisFormatter::class, function () {
```
Comment (oaklees) [inline src/KinesisMonologServiceProvider.php]: We don't really need a `singleton` here as Monolog stores the formatter once retrieved, so `bind` will suffice

File: src/Monolog/KinesisHandler.php [hunk 20]
```diff
@@ -26,27 +25,18 @@ class KinesisHandler extends AbstractProcessingHandler
...
      * KinesisHandler constructor.
      *
-     * @param KinesisClient $client
      * @param string $streamName
-     * @param int $level
+     * @param string $level
      * @param bool $bubble
      */
-    public function __construct(KinesisClient $client, string $streamName, int $level = Logger::INFO, bool $bubble = true)
+    public function __construct(string $streamName, string $level, bool $bubble = true)
```
Comment (oaklees) [inline src/Monolog/KinesisHandler.php]: Can we change `streamName` to `stream` - overly verbose IMO

File: src/Monolog/KinesisHandler.php [hunk 21]
```diff
@@ -26,27 +25,18 @@ class KinesisHandler extends AbstractProcessingHandler
...
     {
         parent::__construct($level, $bubble);
 
-        $this->client = $client;
+        $this->client = app(KinesisClient::class);
         $this->streamName = $streamName;
```
Comment (oaklees) [inline src/Monolog/KinesisHandler.php]: `stream` here too instead

File: src/Monolog/KinesisHandler.php [hunk 22]
```diff
@@ -26,27 +25,18 @@ class KinesisHandler extends AbstractProcessingHandler
...
 
-    /**
-     * Gets the default formatter.
-     *
-     * @return FormatterInterface
-     */
-    protected function getDefaultFormatter()
-    {
-        return new KinesisFormatter();
+        $this->formatter = app(KinesisFormatter::class);
```
Comment (oaklees) [inline src/Monolog/KinesisHandler.php]: Can we move this from the constructor and back to the `getDefaultFormatter` method - the `AbstractProcessingHandler` -> `FormattableHandlerTrait` exposes this for the use case of setting the formatter for a handler

File: tests/Unit/KinesisClientTest.php [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+        $mocked = $this->getMockedKinesisClient();
+
+        $mocked->shouldNotReceive('putRecord');
+
+        $this->app->instance(KinesisClient::class, $mocked);
+
+        logger()->debug('Test debug message');
+    }
+
+    public function testLoggerLogsAboveDefaultInfoLevel()
```
Comment (oaklees) [inline tests/Unit/KinesisClientTest.php]: This should read `testLoggerLogsGreaterThanOrEqualToDefaultInfoLevel`

File: tests/Unit/KinesisClientTest.php [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+    public function loggerLevelTestProvider(): array
+    {
+        return [
+            ['debug'],
+            ['info'],
+            ['notice'],
+            ['warning'],
+            ['error'],
+            ['critical'],
+            ['alert'],
```
Comment (oaklees) [inline tests/Unit/KinesisClientTest.php:24]: Why no `emergency`?

File: tests/Unit/KinesisClientTest.php [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+        $this->app->instance(KinesisClient::class, $mocked);
+
+        logger()->info('Test info message');
+        logger()->warning('Test warning message');
+    }
+
+    /**
+     * @dataProvider loggerLevelTestProvider $logLevel
+     */
+    public function testLoggerDoesNotLogBelowMinimumLevel($logLevel)
```
Comment (oaklees) [inline tests/Unit/KinesisClientTest.php]: I think the dataprovider is overkill here - we could probably just use the same methodology as above and use `twice` on the mock expectation and send two below, and one at the same level

**Review by oaklees — CHANGES_REQUESTED** [review]
Awesome stuff - looking really good - just a couple of notes around some minor tweaks and other stylistic changes. 

Also, some StyleCI formatting to be corrected too.

File: phpunit.xml (hunk 16 above)
Comment (willthorpe) [reply phpunit.xml]: I have no idea what this is :D Will reset back to what it was

**Review by oaklees — APPROVED** [review]
Good stuff  🙌🏻
